lib/Lex/Lexer.cpp

```diff
@@ -0,0 +1,252 @@
+#include "nixf/Lex/Lexer.h"
+#include "nixf/Basic/DiagnosticEngine.h"
+#include "nixf/Syntax/Token.h"
+#include "nixf/Syntax/Trivia.h"
+
+#include <cassert>
+#include <cctype>
+
+namespace nixf {
+
+using namespace tok;
+
+using DK = Diagnostic::DiagnosticKind;
+using NK = Note::NoteKind;
+
+bool Lexer::consumePrefix(std::string_view Prefix) {
+  if (prefix(Prefix)) {
+    Cur += Prefix.length();
+    return true;
+  }
+  return false;
+}
+
+bool Lexer::prefix(std::string_view Prefix) {
+  // Check [Cur, Cur + Prefix.length)
+  if (Cur + Prefix.length() > Src.end())
+    return false;
+  if (remain().starts_with(Prefix)) {
+    return true;
+  }
+  return false;
+}
+
+std::optional<TriviaPiece> Lexer::tryConsumeWhitespaces() {
+  if (eof() || !std::isspace(*Cur))
+    return std::nullopt;
+
+  const char *Ch = Cur;
+  TriviaKind Kind = spaceTriviaKind(*Cur);
+
+  // consume same characters and combine them into a TriviaPiece
+  do {
+    Cur++;
+  } while (Cur != Src.end() && *Cur == *Ch);
+
+  return TriviaPiece(Kind, std::string(Ch, Cur));
+}
+
+std::optional<TriviaPiece> Lexer::tryConsumeComments() {
+  if (eof())
+    return std::nullopt;
+
+  std::string_view Remain = remain();
+  const char *BeginPtr = Cur;
+  unsigned BeginOffset = curOffset();
+
+  if (consumePrefix("/*")) {
+    // consume block comments until we meet '*/'
+    constexpr auto Kind = TriviaKind::BlockComment;
+    while (true) {
+      if (eof()) {
+        // there is no '*/' to terminate comments
+        OffsetRange R = {curOffset() - 1, curOffset()};
+        OffsetRange B = {BeginOffset, BeginOffset + 2};
+
+        Diags.diag(DK::DK_UnterminatedBComment, R)
+            .note(NK::NK_BCommentBegin, B);
+
+        // recover
+        return TriviaPiece(Kind, std::string(Remain));
+      }
+      if (!eof(Cur + 1) && consumePrefix("*/"))
+        // we found the ending '*/'
+        return TriviaPiece(Kind, std::string(BeginPtr, Cur));
+      Cur++;
+    }
+  } else if (consumePrefix("#")) {
+    // single line comments, consume blocks until we meet EOF or '\n' or '\r'
+    while (true) {
+      if (eof() || consumeEOL()) {
+        return TriviaPiece(TriviaKind::LineComment, std::string(BeginPtr, Cur));
+      }
+      Cur++;
+    }
+  } else {
```

Comment:
maybe remove this "else" to ensure we return something on all control flow paths.